pkg/proxy/hybrid/proxy.go

```diff
@@ -7,157 +7,248 @@ import (
-			Namespace: endpoint.Namespace,
-			Name:      endpoint.Name,
-		}
-		if _, ok := p.usingUserspace[svcName]; !ok {
-			forIPTables = append(forIPTables, endpoint)
-		}
+	svcName := types.NamespacedName{
+		Namespace: endpoints.Namespace,
+		Name:      endpoints.Name,
+	}
+
+	wasUsingUserspace, knownEndpoints := p.usingUserspace[svcName]
+	p.usingUserspace[svcName] = p.shouldEndpointsUseUserspace(endpoints)
+
+	if !p.usingUserspace[svcName] {
+		p.mainEndpointsHandler.OnEndpointsAdd(endpoints)
+	}
+
+	// a service could appear before endpoints, so we have to treat this as a potential
+	// state modification for services, and not just an addition (since we could flip proxies).
+	if knownEndpoints && wasUsingUserspace != p.usingUserspace[svcName] {
+		p.switchService(svcName)
 	}
-	return forIPTables
 }
 
-func (p *HybridProxier) OnEndpointsUpdate(endpoints []*api.Endpoints) {
-	p.updateUsingUserspace(endpoints)
+func (p *HybridProxier) OnEndpointsUpdate(oldEndpoints, endpoints *api.Endpoints) {
+	// we track all endpoints in the unidling endpoints handler so that we can succesfully
+	// detect when a service become unidling
+	p.unidlingEndpointsHandler.OnEndpointsUpdate(oldEndpoints, endpoints)
 
-	p.unidlingLoadBalancer.OnEndpointsUpdate(endpoints)
+	p.usingUserspaceLock.Lock()
+	defer p.usingUserspaceLock.Unlock()
 
-	forIPTables := p.getIPTablesEndpoints(endpoints)
-	p.mainEndpointsHandler.OnEndpointsUpdate(forIPTables)
+	svcName := types.NamespacedName{
+		Namespace: endpoints.Namespace,
+		Name:      endpoints.Name,
+	}
 
-	services, err := p.serviceLister.List(labels.Everything())
-	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("Error while listing services from cache: %v", err))
+	wasUsingUserspace, knownEndpoints := p.usingUserspace[svcName]
+	p.usingUserspace[svcName] = p.shouldEndpointsUseUserspace(endpoints)
+
+	if !knownEndpoints {
+		utilruntime.HandleError(fmt.Errorf("received update for unknown endpoints %s", svcName.String()))
+		return
+	}
+
+	isSwitch := wasUsingUserspace != p.usingUserspace[svcName]
+
+	if !isSwitch && !p.usingUserspace[svcName] {
+		p.mainEndpointsHandler.OnEndpointsUpdate(oldEndpoints, endpoints)
 		return
 	}
-	if services == nil {
-		services = []*api.Service{}
+
+	if p.usingUserspace[svcName] {
+		p.mainEndpointsHandler.OnEndpointsDelete(oldEndpoints)
+	} else {
+		p.mainEndpointsHandler.OnEndpointsAdd(endpoints)
 	}
-	p.OnServiceUpdate(services)
+
+	p.switchService(svcName)
+}
+
+func (p *HybridProxier) OnEndpointsDelete(endpoints *api.Endpoints) {
+	// we track all endpoints in the unidling endpoints handler so that we can succesfully
+	// detect when a service become unidling
+	p.unidlingEndpointsHandler.OnEndpointsDelete(endpoints)
+
+	p.usingUserspaceLock.Lock()
+	defer p.usingUserspaceLock.Unlock()
+
+	svcName := types.NamespacedName{
+		Namespace: endpoints.Namespace,
+		Name:      endpoints.Name,
+	}
+
+	usingUserspace, knownEndpoints := p.usingUserspace[svcName]
+
+	if !knownEndpoints {
```

Comment:
Any chance we could get endpoints before a service shows up, just because perhaps we have different watches for each and one watch is stalled in a goroutine somewhere but the other keeps going?